doc/developer/design/20240311_adapter_catalog_listen.md

```diff
@@ -0,0 +1,178 @@
+        self.apply_updates(updates);
+    }
+
+    /// Update in-memory catalog state and generate builtin table updates and controller commands. 
+    fn apply_updates(&mut self, updates: Vec<StateUpdate>) -> (Vec<BuiltinTableUpdate>, Vec<ControllerCommand>) {
+        let mut builtin_table_updates = Vec::new();
+        let mut controller_commands = Vec::new();
+        for StateUpdate { kind, diff } in updates {
+            match (kind, diff) {
+                // ...
+            }
+        }
+        (builtin_table_updates, controller_commands)
+    }
+}
+```
+
+### Coordinator API
+
+```Rust
+impl Coordinator {
+    /// Appends updates to builtin tables.
+    async fn apply_builtin_table_updates(builtin_table_updates: Vec<BuiltinTableUpdate>) {
+        // ...
+    }
+
+    /// Sends commands to the controller.
+    async fn apply_controller_commands(controller_commands: Vec<ControllerCommand>) {
+        // ...
+    }
+}
+```
+
+### Coordinator Bootstrap Psuedo Code
+
+Note: This incorrectly simplifies the current structure of the code and combines some logic from
+`mz_adapter::coord::serve`, `Catalog::open`, and `Coordinator::bootstrap` for ease of reading.
+
+```Rust
+impl Coordinator {
+    async fn serve(&mut self) {
+        // ... 
+        let durable_catalog_state = self.open_durable_catalog_state().await;
+        let catalog = Catalog::new(durable_catalog_state);
+        // Get retractions of current builtin table contents.  
+        let mut builtin_table_updates = self.get_retractions_for_all_builtin_table_contents();
+        let (new_builtin_table_updates, controller_commands) = catalog.sync().await;
+        builtin_table_updates.extend(new_builtin_table_updates);
+        self.apply_builtin_table_updates(builtin_table_updates).await;
+        self.apply_controller_commands(controller_commands).await;
+        // ...
+    }
+}
+```
+
+### Coordinator Sequence Psuedo Code
+
+```Rust
+impl Coordinator {
+    pub(crate) async fn sequence_plan(
+        &mut self,
+        mut ctx: ExecuteContext,
+        plan: Plan,
+        resolved_ids: ResolvedIds,
+    ) -> LocalBoxFuture<'_, ()> {
+        // ...
+        // IMPORTANT: This may trigger a catalog transaction which only updates the durable catalog.
+        self.sequence_X(plan);
+        // ...
+        let (builtin_table_updates, controller_commands) = self.catalog.sync().await;
+        self.apply_builtin_table_updates(builtin_table_updates).await;
+        self.apply_controller_commands(controller_commands).await;
+    }
+}
+```
+
+### Fencing
+
+Fencing will not change as part of this work. That means that a new writer will fence out all
+existing readers and writers.
+
+## Open questions
+
+- Is this a good incremental step towards multi-subscriber or should we go right to
+  multi-subscriber?
+- Should `StateUpdate`s contain timestamps? I.e., while consuming updates from the durable catalog,
+  should the in-memory catalog care at all about timestamps?  
```

Comment:
I think this question here is the key to answering "how do multiple coordinators actually subscribe to the catalog at the same time?", but seems like something we can tackle down the road.